hirefire/contrib/django/middleware.py

```diff
@@ -86,3 +90,42 @@ def process_request(self, request):
 
         elif self.info_path.match(path):
             return self.info(request)
+
+
+class QueueTimeMiddleware(MiddlewareMixin):
+    """
+    The Django middleware that outputs Heroku request queue times to stdout
+
+    The behavior and format for request queue time scaling are described in
+    HireFire's docs: https://help.hirefire.io/article/49-logplex-queue-time
+    """
+    def process_request(self, request):
+        """
+        Outputs the Heroku request queue time to stdout
+        """
+        request_start_header_value = request.META.get("HTTP_X_REQUEST_START")
+        if request_start_header_value is not None:
+            try:
+                request_start_timestamp_ms = int(request_start_header_value)
+            except ValueError:
+                logger.warning(
+                    'Received an invalid "X-Request-Start" header value from '
+                    'Heroku: "%s"',
+                    request_start_header_value,
+                )
+            else:
+                # The timestamp generated by timestamp() is the number of
+                # seconds since the Unix epoch, so it needs to be multiplied
+                # and truncated as an integer to convert it to milliseconds
+                now_timestamp_ms = int(timezone.now().timestamp() * 1000)
+
+                # There may be some clock drift between the Heroku router and
+                # the dyno that is running this middleware; if the calculated
+                # queue time is negative, treat it as zero instead
+                queue_time_ms = (
+                    now_timestamp_ms - request_start_timestamp_ms
+                    if now_timestamp_ms >= request_start_timestamp_ms
+                    else 0
+                )
```

Comment:
Does HireFire not understand clock drift? It seems pretty apparent to me that this is throwing away data, and I hate to see that happen.

Reply:
No, I believe that, unfortunately, the HireFire service will happily ingest a negative queue time value, which affects the aggregate. I think that's less of an issue when one is using a percentile for scaling (e.g. we use 95th percentile queue time in production), but if one uses an average queue time, it could potentially produce an average queue time that is a negative value. Since HireFire has a minimum value of 1ms for the up/down threshold settings, that seems a bit problematic.

Arguably some smarts could be implemented here to add an offset equal to the absolute value of the smallest negative value it has encountered to all future timestamps from the current dyno, but I had hoped to keep things simple. I don't think there's a perfect solution though, since there's always the possibility that the system clock on a dyno will be off compared to the clock on the Heroku router. Anecdotally, however, I've noticed the clock difference between the Heroku router and any given dyno to remain _generally_ within +/-10ms.

All that said, I'm open to suggestions.

Reply:
It's my suggestion that the times be passed along exactly as calculated by default. However, I'd be fine with an option to round off, as long as there's a documentation note about what it does and the trade-offs associated with it.

Reply:
Does that sound alright to you? I just hate to have a default of losing data, even if the data is less than ideal.